text/0035-opentelemetry-protocol.md

```diff
@@ -0,0 +1,425 @@
+- [Open Questions](#open-questions)
+- [Appendix A - Protocol Buffer Definitions](#appendix-a---protocol-buffer-definitions)
+- [Appendix B - Performance Benchmarks](#appendix-b---performance-benchmarks)
+- [Glossary](#glossary)
+- [Acknowledgements](#acknowledgements)
+- [Author's Address](#authors-address)
+
+
+## Motivation
+
+OTLP is a general-purpose telemetry data delivery protocol designed in the scope of OpenTelemetry project. It is an incremental improvement of OpenCensus protocol. Compared to OpenCensus protocol OTLP has the following improvements:
+
+- Ensures high reliability of data delivery and clear visibility when the data cannot be delivered. OTLP uses acknowledgements to implement reliable delivery.
+
+- It is friendly to Level 7 Load Balancers and allows them to correctly map imbalanced incoming traffic to a balanced outgoing traffic. This allows to efficiently operate large networks of nodes where telemetry data generation rates change over time.
+
+- Allows backpressure signalling from telemetry data destinations to sources. This is important for implementing reliable multi-hop telemetry data delivery all the way from the source to the destination via intermediate nodes, each having different processing capacity and thus requiring different data transfer rates.
+
+## Protocol Details
+
+OTLP defines the encoding of telemetry data and the protocol used to exchange data between the client and the server.
+
+This specification defines how OTLP is implemented over [gRPC](https://grpc.io/) and specifies corresponding [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview) schema. Future extensions to OTLP may define implementations over other transports. For details of gRPC service definition see section [gRPC Transport](#grpc-transport).
+
+OTLP is a request/response style protocols: the clients send requests, the server replies with corresponding responses. This document defines one requests and response type: `Export`.
+
+### Export Request and Response
+
+After establishing the underlying transport the client starts sending telemetry data using `Export` requests.  The client continuously sends a sequence of `Export` requests to the server and expects to receive a response to each request:
+
+![Request-Response](images/otlp-request-response.png)
+
+_Note: this protocol is concerned with reliability of delivery between one pair of client/server nodes and aims to ensure that no data is lost in-transit between the client and the server. Many telemetry collection systems have intermediary nodes that the data must travel across until reaching the final destination (e.g. application -> agent -> collector -> backend). End-to-end delivery guarantees in such systems is outside of the scope of OTLP. The acknowledgements described in this protocol happen between a single client/server pair and do not span intermediary nodes in multi-hop delivery paths._
```

Comment:
Perhaps worth clarifying that it doesn't mean that those chained acks cannot happen, eg if the "server" is just an agent/sidecar that enriches/translates messages and forwards them further, can it respond to the client based on whether the downstream collector acked the message?

Reply:
I'd widen this - does "ack" means delivered and saved in any sort of persistence storage or can be replied when messages saved in-memory and can be lost on node restart. Or not defined?

Reply:
@yurishkuro Technically yes, nothing prevents the intermediary node to only ack after downstream acked.
It is a choice of particular _usage_ of the protocol to define what it means that the data is acknowledged. It may mean that:
- The data is received.
- The data is received and successfully decoded.
- The data is received, decoded and persisted.
- The data is received and forwarded.
etc.
The protocol does not define this, that definition belongs to the abstraction level which uses the protocol. (@SergeyKanzhelev I believe this answers your question too).

The only reason I have this note is to make it clear that acknowledgement response travels between a single connected gRPC client/server pair and OTLP does not attempt to implement end-to-end reliable data delivery on its own (but it certainly can be used to achieve that).